bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java

```diff
@@ -503,17 +513,18 @@ public void run() {
                     // responses
                     for (int i = 0; i < requestsCount; i++) {
                         ForceWriteRequest req = localRequests.get(i);
-                        numReqInLastForceWrite += req.process();
+                        numReqInLastForceWrite += req.process(writeHandlers);
                         req.recycle();
                     }
 
                     journalStats.getForceWriteGroupingCountStats()
                             .registerSuccessfulValue(numReqInLastForceWrite);
-
-                    if (requestProcessor != null) {
-                        requestProcessor.flushPendingResponses();
+                    if (writeHandlers.size() > 0) {
+                        for (ObjectCursor<BookieRequestHandler> writeHandler : writeHandlers) {
+                            writeHandler.value.flushPendingResponse();
+                        }
+                        writeHandlers.clear();
                     }
```

Comment:
In this case it would efficient to use the lambda approach because it would be non-capturing and `ObjectHashSet` is not going to create an iterator either.

```suggestion
                    writeHandlers.forEach(wh -> wh.flushPendingResponse());
                    writeHandlers.clear();
```